# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 5: While Loops, Conditional Statements, IRR and YTM

The ABC Company has a mature project with declining revenues and increasing costs. Net Cashflows are projected as follows:

- 500 in year 1,
- 350 in year 2, 
- 200 in year 3,
- 50 in year 4,
- -100 in year 5,
- -200 in year 6

1. Separate __positive__ and __negative__ cashflows. __Create two new lists__!

In [1]:
#run the cell
cf = [500, 350, 200, 50, -100, -200]

In [2]:
cf_pos =[]
cf_neg =[]

for i in cf: #complete the code!
    if i > 0:
        cf_pos.append(i)
    elif i < 0:
        cf_neg.append(i)
print(cf_pos)
print(cf_neg)

[500, 350, 200, 50]
[-100, -200]


2. Get the __first unprofitable year__. Your Code should return the statement __"Year ... is the first unprofitable Year!"__ 

In [6]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the first unprofitable year".format(n))
        break
    n += 1

Year 5 is the first unprofitable year


3. Get the __last profitable year__. Your Code should return the statement __"Year ... is the last profitable Year!"__ 

In [8]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the last proftiable year".format(n -1))
        break
    n += 1

Year 4 is the last proftiable year


The __modulo operator %__ returns the __remainder__ of the division between two numbers. For instance, 5 % 3 returns 2. 

In [26]:
# run the cell
5 % 3

2

In [27]:
# run the cell
14 % 5

4

If a is divisible by b __without remainder__, a % b __returns 0__. 

In [28]:
# run the cell
a = 9
b = 3
a % b

0

4. Return the __first number n__ that is __divisible by 362 and 216__ (without remainder).<br>
Your code shall return the following statement: __"... is the first number divisible by 362 and 216!"__

In [9]:
a = 362
b = 216

n = 1 
while True: #complete the code!
    
    if n % a == 0 and n % b == 0:
        print("{} is the first number divisible by {} and {}!".format(n, a, b))
        break
    n += 1

39096 is the first number divisible by 362 and 216!


(result: 39096)

The ABC Company evaluates to buy an additional machine that will increase profits/cashflows by
- 50 USD in year 1,
- 75 USD in year 2 ,
- 100 USD in year 3, 
- 150 USD in year 4,
- 125 USD in year 5. (each cf at period´s end)

The machine costs 450 USD (Investment in to). 

5. Calculate the __Project´s IRR__ and evaluate whether ABC should pursue the project. ABCs required rate of return is 6% p.a.

In [16]:
cf = [-450, 50, 75, 100, 150, 125]
guess = 0.06

In [17]:
f = 1 + guess
step = 0.000001
target_npv = 0
tolerance = 0.001
npv = 0

while True:
    f = 1 + guess
    npv = 0
    for i in range(len(cf)):
        npv += cf[i] / f**(i) #simple nvp calc
    diff = npv - target_npv #diff shows how far the npv we're calcing is from the target npv (0)
    
    #iterates over the difference in npv to get the npv within the tolerance range
    if abs(diff) > tolerance:
        if diff < 0: #if difference is negative, then decrease the step - to get it in the tolerance range
            guess -= step
        elif diff > 0: #if difference is positive, then increase the step - to get it in the tolerance range
            guess += step
    else:
        break

print(npv, guess)

-0.0002819830025373449 0.03124499999997124


(result: 3.12%)

5. Would your __decision change__ if you can increse year 5 cashflows from __125 to 175__? 

In [18]:
cf[-1] = 175

In [19]:
f = 1 + guess
step = 0.000001
target_npv = 0
tolerance = 0.001
npv = 0

while True:
    f = 1 + guess
    npv = 0
    for i in range(len(cf)):
        npv += cf[i] / f**(i) #simple nvp calc
    diff = npv - target_npv #diff shows how far the npv we're calcing is from the target npv (0)
    
    #iterates over the difference in npv to get the npv within the tolerance range
    if abs(diff) > tolerance:
        if diff < 0: #if difference is negative, then decrease the step - to get it in the tolerance range
            guess -= step
        elif diff > 0: #if difference is positive, then increase the step - to get it in the tolerance range
            guess += step
    else:
        break

print(npv, guess)

-0.0003708748203337109 0.058286999999998285


(result: 5.83%)

6. The ABC Government issued a 30Y Government Bond one year ago with a Coupon Rate of 5.5% (annual payments in arrears). Today´s Bond Price is 97.2 (per 100 par Value). Calculate the Bond´s __current YTM__.

In [2]:
#add 20 payments to the string
cf = [5.5] * 29

#add the final repayment of the pricipal
cf[-1] = cf[-1] + 100 

#insert the price paid for the bond at t=0
cf.insert(0, -97.2)

#print the cashflows
cf

[-97.2,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 5.5,
 105.5]

In [5]:
guess = 0.06
step = 0.000001
target_npv = 0
tolerance = 0.001

while True:
    f= 1 + guess
    npv = 0
    for i in range(len(cf)):
        npv += cf[i] / f**(i)
    diff = npv - target_npv
        
    #iterates over the difference in npv to get the npv within the tolerance range
    if abs(diff) > tolerance:
        if diff < 0: #if difference is negative, then decrease the step - to get it in the tolerance range
            guess -= step
        elif diff > 0: #if difference is positive, then increase the step - to get it in the tolerance range
            guess += step
    else:
        break
        
print(npv, guess)

-0.00023920511386066323 0.056995999999996994


(result: 5.70%)

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

The ABC Company has a mature project with declining revenues and increasing costs. Net Cashflows are projected as follows:

- 500 in year 1,
- 350 in year 2, 
- 200 in year 3,
- 50 in year 4,
- -100 in year 5,
- -200 in year 6

1. Separate __positive__ and __negative__ cashflows. __Create two new lists__!

In [ ]:
#run the cell
cf = [500, 350, 200, 50, -100, -200]

In [ ]:
cf_pos =[]
cf_neg =[]

for i in cf:
    if i > 0:
        cf_pos.append(i)
    elif i < 0:
        cf_neg.append(i)
print(cf_pos, cf_neg)

2. Get the __first unprofitable year__. Your Code should return the statement __"Year ... is the first unprofitable Year!"__ 

In [ ]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the first unprofitable year!".format(n))
        break
    n += 1

3. Get the __last profitable year__. Your Code should return the statement __"Year ... is the last profitable Year!"__ 

In [ ]:
n = 1
for i in cf:
    if i < 0:
        print("Year {} is the last profitable year!".format(n - 1))
        break
    n += 1

The __modulo operator %__ returns the __remainder__ of the division between two numbers. For instance, 5 % 3 returns 2. 

In [ ]:
# run the cell
5 % 3

In [ ]:
# run the cell
14 % 5

If a is divisible by b __without remainder__, a % b __returns 0__. 

In [ ]:
# run the cell
a = 9
b = 3
a % b

4. Return the __first number n__ that is __divisible by 362 and 216__ (without remainder).<br>
Your code shall return the following statement: __"... is the first number divisible by 362 and 216!"__

In [ ]:
a = 362
b = 216

n = 1
while True:
    if (n%a == 0) and (n%b == 0):
        print("{} is the first number divisible by {} and {}!".format(n, a, b))
        break
    n+=1

(result: 39096)

The ABC Company evaluates to buy an additional machine that will increase profits/cashflows by
- 50 USD in year 1,
- 75 USD in year 2 ,
- 100 USD in year 3, 
- 150 USD in year 4,
- 125 USD in year 5. (each cf at period´s end)

The machine costs 450 USD (Investment in to). 

5. Calculate the __Project´s IRR__ and evaluate whether ABC should pursue the project. ABCs required rate of return is 6% p.a.

In [ ]:
cf = [-450, 50, 75, 100, 150, 125]

In [ ]:
guess = 0.06
step = 0.000001
target_npv = 0
tolerance = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tolerance:
        if diff < 0:
            guess -= step
        elif diff > 0:
            guess += step
    else:
        break
        
print(NPV, guess) 

(result: 3.12% -> reject)

5. Would your __decision change__ if you can increse year 5 cashflows from __125 to 175__? 

In [ ]:
cf[-1] = 175

In [ ]:
guess = 0.06
step = 0.0000001
target_npv = 0
tolerance = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tolerance:
        if diff < 0:
            guess -= step
        elif diff > 0:
            guess += step
    else:
        break
        
print(NPV, guess) 

(result: 5.83% -> still reject)

6. The ABC Government issued a 30Y Government Bond one year ago with a Coupon Rate of 5.5% (annual payments in arrears). Today´s Bond Price is 97.2 (per 100 par Value). Calculate the Bond´s __current YTM__.

In [ ]:
cf = [5.5] * 29

In [ ]:
cf[-1] = cf[-1] + 100

In [ ]:
cf.insert(0, -97.2)

In [ ]:
cf

In [ ]:
guess = 0.06
step = 0.0000001
target_npv = 0
tolerance = 0.001 

while True:
    f = 1 + guess
    NPV = 0
    for i in range(len(cf)):
        NPV += cf[i] / f**(i)
    diff = NPV - target_npv
    
    if abs(diff) > tolerance:
        if diff < 0:
            guess -= step
        elif diff > 0:
            guess += step
    else:
        break
        
print(NPV, guess) 

(result: 5.70%)